ARO-17124 allow changing cluster identity

yithian opened the pull request:
### Which issue this PR addresses:

<!--
Please include a link to the ADO work item as well as any GitHub issues.

Usage: `Fixes #<GitHub issue number>`, or `Fixes (paste link of issue)`.
-->
Fixes ARO-17124

### What this PR does / why we need it:

<!--
Include a brief summary of what the PR is intended to accomplish and how the PR
does it. (2-3 sentences)
-->

This PR adds functionality to the RP and the CLI that allows the customer to change the identity assigned to the cluster

### Test plan for issue:

<!--
How did you test that this PR works?

- Are there unit tests?
- Are there integration/e2e tests?
- If it is not possible to write automated tests, explain why and document how
  to manually test and verify the feature.
-->

Included a unit test for the updated putorpatch code and manually tested against a localdev RP with a dev cluster. After running the update via the CLI I verified in CosmosDB that the cluster's identity has been updated

### Is there any documentation that needs to be updated for this PR?

<!--
- If yes and the docs are in GitHub, include doc updates in the PR.
- If yes and the docs are not in GitHub (i.e. ADO wiki), include a link to the
  docs.
- If no, explain why (e.g. "tech debt cleanup, N/A").
-->

Maybe? This may need to be documented in customer-facing documentation for ARO

### How do you know this will function as expected in production? 

<!--
- Does adequate telemetry, monitoring and documentation exist to effectively operate your change?
- Have failure modes been identified and mitigated? 
-->

This has been tested in localdev and will be tested again in canary

kimorris27 reviewed (commented):
Looks great! Just a few small suggestions and questions.

yithian commented on pkg/api/v20240812preview/openshiftcluster_validatestatic.go:
```diff
@@ -516,10 +516,6 @@ func (sv openShiftClusterStaticValidator) validatePlatformIdentities(oc *OpenShi
 		return api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeInvalidParameter, "identity", "Cluster identity and platform workload identities require each other.")
 	}
 
-	if clusterIdentityPresent && len(oc.Identity.UserAssignedIdentities) != 1 {
```
When testing the CLI changes, I was getting an error from this validation, I think because [the number of UserAssignedIdentities is briefly 0](https://github.com/Azure/ARO-RP/pull/4272/files#diff-c0dceb05e2a59300f2e11380f3601ea682171f3f4837bb769689fc79e78b1a70R231). I'll re-add the validation locally and confirm that the CLI is still tripping over it, since I ran into that issue before making some other changes to the RP logic

yithian commented:
@cadenmarchese @kimorris27 I think I've addressed your feedback and concerns and have manually confirmed that these changes successfully update the identity for the cluster. Would you mind re-reviewing?

kimorris27 reviewed (commented):
I gave another review, and I have nothing to add except that I agree with both of Caden's suggestions.

yithian commented:
Canary testing was successful

```
[achvatal@xoth ~]$ az aro list -g achvtest2 -o json | jq -r ".[0].identity.userAssignedIdentities"
{
  "/subscriptions/60bf318d-6914-4105-a3b5-d0d2c10388c8/resourceGroups/achvtest2/providers/Microsoft.ManagedIdentity/userAssignedIdentities/aro-Cluster": {
    "clientId": "4fd93196-c9c4-4a19-961f-9ec019275912",
    "principalId": "10f9dae8-9a58-49e9-8478-b63bcb6e66ed"
  }
}
[achvatal@xoth ~]$ az aro update -g achvtest2 -n achvtest --assign-cluster-identity /subscriptions/60bf318d-6914-4105-a3b5-d0d2c10388c8/resourcegroups/achvtest2/providers/Microsoft.ManagedIdentity/userAssignedIdentities/aro-Cluster2
Argument '--assign-cluster-identity' is in preview and under development. Reference and support levels: https://aka.ms/CLI_refstatus
Location     Name      ProvisioningState    ResourceGroup
-----------  --------  -------------------  ---------------
eastus2euap  achvtest  Succeeded            achvtest2
[achvatal@xoth ~]$ az aro list -g achvtest2 -o json | jq -r ".[0].identity.userAssignedIdentities"
{
  "/subscriptions/60bf318d-6914-4105-a3b5-d0d2c10388c8/resourcegroups/achvtest2/providers/Microsoft.ManagedIdentity/userAssignedIdentities/aro-Cluster2": {
    "clientId": "154b4393-1557-4d06-8287-681f61456fc2",
    "principalId": "f9f35efb-91e5-4326-b421-42fce12486c2"
  }
}```

cadenmarchese commented:
/azp run ci

shubhadapaithankar reviewed (approved):
lgtm
